tests/functional/test_s3.py

```diff
@@ -148,6 +149,34 @@ def test_500_error_with_non_xml_body(self):
         self.assertEqual(response['ResponseMetadata']['HTTPStatusCode'], 200)
 
 
+class TestS3SigV4(BaseS3OperationTest):
+    def setUp(self):
+        super(TestS3SigV4, self).setUp()
+        self.client = self.session.create_client(
+            's3', self.region, config=Config(signature_version='s3v4'))
+        self.response_mock = mock.Mock()
+        self.response_mock.content = b''
+        self.response_mock.headers = {}
+        self.response_mock.status_code = 200
+        self.http_session_send_mock.return_value = self.response_mock
+
+    def get_sent_headers(self):
+        return self.http_session_send_mock.mock_calls[0][1][0].headers
+
+    def test_content_md5_set(self):
+        self.client.put_object(Bucket='foo', Key='bar', Body='baz')
+        self.assertIn('content-md5', self.get_sent_headers())
+
+    def test_content_sha256_set_if_md5_is_unavailable(self):
+        with mock.patch('botocore.auth.MD5_AVAILABLE', False):
+            with mock.patch('botocore.handlers.MD5_AVAILABLE', False):
+                self.client.put_object(Bucket='foo', Key='bar', Body='baz')
+        sent_headers = self.get_sent_headers()
+        unsigned = 'UNSIGNED-PAYLOAD'
+        self.assertNotEqual(sent_headers['x-amz-content-sha256'], unsigned)
+        self.assertNotIn('content-md5', sent_headers)
+
```

Comment:
Should we add a test for when the config value of opting-in to sigv4 signing is turned on?